File: java/src/main/java/com/gpayments/requestor/testlab/threeds1/ThreeDS1Service.java
package com.gpayments.requestor.testlab.threeds1;

import com.gpayments.requestor.testlab.AuthServiceV2;
import com.gpayments.requestor.testlab.config.Config;
import com.gpayments.requestor.testlab.dto.threeds1.ThreeDS1AuthReq;
import com.gpayments.requestor.testlab.dto.threeds1.ThreeDS1AuthResp;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

/**
 * 3DSecure 1.0.2 service class
 *
 * @author GPayments
 */
@Service
public class ThreeDS1Service {

  private static final Logger logger = LoggerFactory.getLogger(ThreeDS1Service.class);

  private final RestTemplate restTemplate;
  private final String authUrl;
  private final Config config;

  @Autowired
  public ThreeDS1Service(RestTemplate restTemplate,
      Config config) {
    this.restTemplate = restTemplate;
    this.authUrl = config.getAsAuthUrl() + "/api/v2/auth/3ds1";
    this.config = config;
  }


  ThreeDS1AuthResp handleAuthRequest(ThreeDS1AuthReq request) {

    //generate the transaction id, this is optional.
    request.setThreeDSRequestorTransID(UUID.randomUUID().toString());

    logger.info("sending 3ds1 auth request to ActiveServer: {}", authUrl);

    HttpEntity<ThreeDS1AuthReq> httpRequest;
    if (config.isGroupAuth()) {
      HttpHeaders headers = new HttpHeaders();
      headers.add(AuthServiceV2.AS_MERCHANT_TOKEN_HEADER, config.getMerchantToken());
      httpRequest = new HttpEntity<>(request, headers);
    } else {

      httpRequest = new HttpEntity<>(request);
    }

    ResponseEntity<ThreeDS1AuthResp> response = restTemplate
        .postForEntity(authUrl, httpRequest, ThreeDS1AuthResp.class);

    if (response.getStatusCode() == HttpStatus.OK) {

      ThreeDS1AuthResp body = response.getBody();
      logger.info("server returns ok, content: {}", body);

      return body;

    } else {
      ThreeDS1AuthResp body = response.getBody();
      logger.error("server returns error code: {}, content: {}", response.getStatusCode(),
          body);
      return body;
    }

  }


}

File: dotnet/Controllers/Threeds1/ThreeDS1Service.cs
using GPayments.Requestor.TestLab.Helpers;
using GPayments.Requestor.TestLab.Models.dto.Threeds1;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http.Headers;
using System.Web;

namespace GPayments.Requestor.TestLab.Controllers.Threeds1
{
    /**
	 * 3DSecure 1.0.2 service class
	 *
	 * @author GPayments
	 */
    public class ThreeDS1Service
    {
        private static readonly log4net.ILog logger = log4net.LogManager.GetLogger(typeof(ThreeDS1Service));
        private string authUrl;

        public ThreeDS1Service()
        {
            this.authUrl = Config.AsAuthUrl + "/api/v2/auth/3ds1";
        }

        public ThreeDS1AuthResp HandleAuthRequest(ThreeDS1AuthReq request)
        {
            //generate the transaction id, this is optional.
            request.threeDSRequestorTransID = Guid.NewGuid().ToString();

            logger.Info(string.Format("sending 3ds1 auth request to ActiveServer: {0}", authUrl));

            ThreeDS1AuthResp response = (ThreeDS1AuthResp)RestClientHelper.PostForObject(authUrl, request, typeof(ThreeDS1AuthResp));

            logger.Info(string.Format("server returns ok, content: {0}", response));

            return response;
        }
    }
}